digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala

```diff
@@ -213,5 +213,81 @@ class ProjectService(roadAddressService: RoadAddressService, roadLinkService: Ro
     }
   }
 
+  def getRoadAddressChangesAndSendToTR(projectId: Set[Long]) = {
+    val roadAddressChanges = RoadAddressChangesDAO.fetchRoadAddressChanges(projectId)
+    ViiteTierekisteriClient.sendRoadAddressChangeData(roadAddressChanges)
+  }
+
+  def getProjectStatusFromTR(projectId: Long) = {
+    ViiteTierekisteriClient.getProjectStatus(projectId.toString)
+  }
+
+
+
+
+  val listOfExitStatuses=List(1,3,5) // closed, errorinTR,savedtotr magic numbers
+
 
+
+  private def getStatusFromTRObject(trProject:Option[trPojectStatus]):Option[ProjectState] = {
+    trProject match {
+      case Some(trPojectobject) => mapTRstateToViiteState(trPojectobject.status.getOrElse(""))
+      case None => None
+      case _ => None
+    }
+  }
+
+  def updateProjectStatusIfNeeded(currentStatus:ProjectState, newStatus:ProjectState,projectid:String) :(ProjectState)= {
+    if (currentStatus.value!=newStatus.value && newStatus.value!=99) //magic numbers
+    {
+      ProjectDAO.updateProjectStatus(projectid,newStatus)
+    }
+    if (newStatus.value!=99){
+      newStatus
+    } else
+    {
+      currentStatus
+    }
+  }
+
+  def updateProjectsWaitingResponseFromTR(): Unit =
+  {
+    withDynSession {
```

Comment:
Use withDynTransaction, use per project basis: not updating all projects at once but locking failures / deadlocks should roll back only that particular project